switchboard/models.py

```diff
@@ -66,172 +80,131 @@ def create(cls, **kwargs):
-            cls.update(result, result, upsert=True)
-            result = cls.c.find_one(kwargs)
-            instance = cls(**result)
+            instance = cls.update(data, data, upsert=True)
         else:
             created = False
-            instance = cls(**result)
         return instance, created
 
     @classmethod
-    def find(cls, **kwargs):
-        return [cls(**s) for s in cls.c.find(kwargs)]
-
-    @classmethod
-    def update(cls, spec, document, upsert=False):
+    def update(cls, spec, updates, upsert=False):
         '''
-        Mimics a subset of PyMongo's Collection.update functionality. The spec
-        is used to search for the document to update, document contains the
+        The spec is used to search for the data to update, updates contains the
         values to be updated, and upsert specifies whether to do an insert if
-        the original document is not found.
+        the original data is not found.
         '''
-        previous = cls.get(**spec)
-        cls.pre_save.send(previous)
-        result = cls.c.update(spec, document, upsert=upsert)
-        current = cls.get(**spec)
-        cls.post_save.send(current)
-        return result
+        if 'key' in spec:
+            previous = cls.get(spec['key'])
+        else:
+            previous = None
+        if previous:
+            # Update existing data.
+            current = cls(**previous.__dict__)
+        elif upsert:
+            # Create new data.
+            current = cls(**spec)
+        else:
+            current = None
+        # XXX Should there be any error thrown if this is a noop?
+        if current:
+            cls.pre_save.send(current)
+            current.__dict__.update(updates)
+            current.save()
+            cls.post_save.send(current)
+        return current
 
     @classmethod
-    def remove(cls, **kwargs):
-        instance = cls.get(**kwargs)
-        cls.pre_delete.send(instance)
-        result = cls.c.remove(kwargs)
-        cls.post_delete.send(instance)
+    def remove(cls, key):
+        key = _key(key)
+        instance = cls.get(key)
+        if instance:
+            cls.pre_delete.send(instance)
+            result = cls.ds.delete(key)
+            cls.post_delete.send(instance)
+        else:
+            # XXX Should there be any error thrown if this is a noop?
+            result = None
         return result
 
     @classmethod
     def all(cls):
-        return [cls(**s) for s in cls.c.find()]
+        if cls._is_redis_datastore():
+            return cls._redis_all()
+        else:
+            query = datastore.Query(_key())
+            return [cls(**data) for data in cls.ds.query(query)]
 
     @classmethod
-    def count(cls):
-        return cls.c.count()
+    def _is_redis_datastore(cls):
+        '''
+        Check to see if the datastore is Redis.
+        '''
+        try:
+            import datastore.redis
+            return isinstance(cls.ds, datastore.redis.RedisDatastore)
```

Comment:
Please move the return at the end, so that only the code that we want to protect for that exception is in the try/except block.